File: src/server/crdt/dot_context.go
package crdt

import (
	"fmt"
	"sdle-server/utils"
)

type DotContext struct {
	compactContext map[string]int
	dots           utils.Set[Dot]
}

func NewDotContext() *DotContext {
	return &DotContext{
		compactContext: make(map[string]int),
		dots:           utils.NewSet[Dot](),
	}
}

func (dc *DotContext) Knows(dot Dot) bool {
	if localSeq, ok := dc.compactContext[dot.id]; ok {
		return dot.seq <= localSeq
	}

	return false
}

func (dc *DotContext) MakeDot(replicaID string) Dot {
	if localSeq, ok := dc.compactContext[replicaID]; ok {
		localSeq++
		dc.compactContext[replicaID] = localSeq
		return NewDot(replicaID, localSeq)
	} else {
		dc.compactContext[replicaID] = 1
		return NewDot(replicaID, 1)
	}
}

func (dc *DotContext) InsertDot(dot Dot) {
	dc.InsertDotCompact(dot, true)
}

func (dc *DotContext) InsertDotCompact(dot Dot, compact bool) {
	dc.dots.Add(dot)
	if compact {
		dc.Compact()
	}
}

func (dc *DotContext) Compact() {
	dotAdded := true
	for dotAdded {
		dotAdded = false

		for dot := range dc.dots {
			if localSeq, ok := dc.compactContext[dot.id]; ok { // Has entry in compact context
				if dot.seq == localSeq + 1 { // Dot is sequentially after, can compact
					dc.compactContext[dot.id] = dot.seq
					dc.dots.Remove(dot)
					dotAdded = true

				} else if dot.seq <= localSeq { // Dot is already included, ignore
					dc.dots.Remove(dot)
				}

			} else { // No entry in compact context exists
				if dot.seq == 1 { // Can compact
					delete(dc.compactContext, dot.id)
					dc.dots.Remove(dot)
					dotAdded = true
				}
			}
		}
	}
}

func (dc *DotContext) Join(other *DotContext) {
	if dc == other {
		return
	}

	for replicaID, otherSeq := range other.compactContext {
		if localSeq, ok := dc.compactContext[replicaID]; ok {
			dc.compactContext[replicaID] = max(localSeq, otherSeq)
		} else {
			dc.compactContext[replicaID] = otherSeq
		}
	}

	for otherDot := range other.dots {
		dc.dots.Add(otherDot)
	}

	dc.Compact()
}

func (dc *DotContext) Clone() *DotContext {
	clone := NewDotContext()

	for replicaID, seq := range dc.compactContext {
		clone.compactContext[replicaID] = seq
	}

	for dot := range dc.dots {
		clone.dots.Add(dot)
	}

	return clone
}

func (dc *DotContext) String() string {
	return fmt.Sprintf("DotContext{compactContext: %v, dots: %v}", dc.compactContext, dc.dots)
}

File: src/client/types/crdt/generic/dot-context.ts
import Dot from "./dot";
import { DotContext as DotContextProto, Dot as DotProto } from "@/lib/proto/global";

export default class DotContext {
    private versionVector: Map<string, number>;
    private dotCloud: Set<string>;  // Using string representation, since objects do not work as Set keys

    constructor() {
        this.versionVector = new Map<string, number>();
        this.dotCloud = new Set<string>();
    }

    public knows(dot: Dot): boolean {
        const localSeq = this.versionVector.get(dot.id) || 0;
        return dot.seq <= localSeq || this.dotCloud.has(dot.toKey());
    }

    public makeDot(id: string): Dot {
        const localSeq = this.versionVector.get(id) || 0;
        const newSeq = localSeq + 1;

        this.versionVector.set(id, newSeq);
        return new Dot(id, newSeq);
    }

    public insertDot(dot: Dot, compact: boolean = true): void {
        this.dotCloud.add(dot.toKey());
        if (compact) {
            this.compact();
        }
    }

    public compact(): void {
        let changed = true;

        while (changed) {
            changed = false;

            this.dotCloud.forEach((dotKey) => {
                const dot = Dot.fromKey(dotKey);
                const localSeq = this.versionVector.get(dot.id) || 0;

                if (dot.seq === localSeq + 1) {  // Sequential order, can be merged into version vector
                    this.versionVector.set(dot.id, dot.seq);
                    this.dotCloud.delete(dotKey);
                    changed = true;

                } else if (dot.seq <= localSeq) {  // Already known, can be removed
                    this.dotCloud.delete(dotKey);
                    changed = true;
                }
            });
        }
    }

    public join(other: DotContext): void {
        // Optimization for complex CRDTs, which frequently join a context with itself
        if (this === other) {
            return;
        }

        // Pointwise maximum of version vectors
        other.versionVector.forEach((otherSeq, id) => {
            const localSeq = this.versionVector.get(id) || 0;
            this.versionVector.set(id, Math.max(localSeq, otherSeq));
        })

        this.dotCloud = this.dotCloud.union(other.dotCloud);
        this.compact();
    }

    public clone(): DotContext {
        const newCtx = new DotContext();

        this.versionVector.forEach((seq, id) => {
            newCtx.versionVector.set(id, seq);
        });

        this.dotCloud.forEach((dot) => {
            newCtx.dotCloud.add(dot);
        });

        return newCtx;
    }

    public copy(other: DotContext): void {
        if (this === other) {
            return;
        }

        this.versionVector = new Map<string, number>(other.versionVector);
        this.dotCloud = new Set<string>(other.dotCloud);
    }

    public toProto(): DotContextProto {
        const protoDots = Array.from(this.dotCloud).map(dotKey => {
            return Dot.fromKey(dotKey).toProto();
        });

        const protoVersionVector: { [key: string]: number } = {};
        this.versionVector.forEach((seq, id) => {
            protoVersionVector[id] = seq;
        });

        return DotContextProto.create({
            versionVector: protoVersionVector,
            dots: protoDots,
        });
    }

    public static fromProto(proto: DotContextProto): DotContext {
        const context = new DotContext();

        for (const id in proto.versionVector) {
            context.versionVector.set(id, proto.versionVector[id]);
        }

        for (const dotProto of proto.dots) {
            const dot = Dot.fromProto(dotProto as DotProto);
            context.dotCloud.add(dot.toKey());
        }

        return context;
    }
}